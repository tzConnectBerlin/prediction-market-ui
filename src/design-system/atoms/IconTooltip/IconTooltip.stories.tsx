import * as React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { IconTooltip, IconTooltipProps } from './IconTooltip';
import { Typography } from '../Typography';

export default {
  title: 'Atom/IconTooltip',
  component: IconTooltip,
} as Meta;

const Template: Story<IconTooltipProps> = (args) => <IconTooltip {...args} />;

export const Auto = Template.bind({});
Auto.args = {
  description: 'simple description',
};

export const ElementDescription = Template.bind({});
ElementDescription.args = {
  description: (
    <div>
      Sale price is discounted by{' '}
      <Typography component="span" fontWeight="bold">
        5%
      </Typography>
      . This fee goes to the market creator and liquidity providers.
    </div>
  ),
  placement: 'bottom',
};

export const LongDescription = Template.bind({});
LongDescription.args = {
  description:
    'Sale price is discounted by 5%. This fee goes to the market creator and liquidity providers.',
  placement: 'bottom-end',
};

export const MaxWidth = Template.bind({});
MaxWidth.args = {
  description:
    'Sale price is discounted by 5%. This fee goes to the market creator and liquidity providers.',
  placement: 'bottom-end',
  maxWidth: '300px',
};
