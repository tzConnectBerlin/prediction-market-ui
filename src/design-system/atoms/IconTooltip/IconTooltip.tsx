import * as React from 'react';
import { IconButton, Paper, Popper, PopperPlacementType, Theme, useTheme } from '@material-ui/core';
import styled from '@emotion/styled';
import { IoMdClose, IoMdInformation } from 'react-icons/io';
import { Typography } from '../Typography';
import { CircleBackground } from '../CloseIcon';

interface StyledPaperProps {
  maxWidth: string;
  theme: Theme;
}

const StyledPaper = styled(Paper)<StyledPaperProps>`
  padding: ${({ theme }) => theme.spacing(3)};
  max-width: ${({ maxWidth }) => maxWidth};
`;

export interface IconTooltipProps {
  description: string | React.ReactNode;
  /**
   * the place of opener
   */
  placement?: PopperPlacementType;
  /**
   * Disable or Enable button
   */
  disabled?: boolean;
  /**
   * icon color
   */
  iconColor?: string;
  /**
   * maximum width of opener
   */
  maxWidth?: string;
}

export const IconTooltip: React.FC<IconTooltipProps> = ({
  disabled,
  description,
  placement = 'auto',
  iconColor,
  maxWidth,
}) => {
  const theme = useTheme();
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const [poperPlacement, setPlacement] = React.useState<PopperPlacementType>();
  const [open, setOpen] = React.useState(false);

  const handleClick =
    (newPlacement: PopperPlacementType) => (event: React.MouseEvent<HTMLButtonElement>) => {
      setAnchorEl(event.currentTarget);
      setOpen((prev) => placement !== newPlacement || !prev);
      setPlacement(newPlacement);
    };

  const id = open ? 'tooltip-popper' : undefined;
  const fillColor = iconColor ?? theme.palette.primary.main;

  return (
    <>
      <IconButton
        type="button"
        aria-label="tooltip"
        onClick={handleClick(placement)}
        disabled={disabled}
      >
        <CircleBackground>
          {open ? (
            <IoMdClose fill={fillColor} size="0.6em" />
          ) : (
            <IoMdInformation fill={fillColor} size="0.7em" />
          )}
        </CircleBackground>
      </IconButton>
      <Popper id={id} open={open} anchorEl={anchorEl} placement={poperPlacement}>
        <StyledPaper maxWidth={maxWidth ?? 'auto'} theme={theme}>
          <Typography size="body2">{description}</Typography>
        </StyledPaper>
      </Popper>
    </>
  );
};
