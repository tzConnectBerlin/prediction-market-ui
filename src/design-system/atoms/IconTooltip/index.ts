export { IconTooltip as default, IconTooltip } from './IconTooltip';
export type { IconTooltipProps } from './IconTooltip';
