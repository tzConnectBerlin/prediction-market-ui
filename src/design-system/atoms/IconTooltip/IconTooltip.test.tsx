import { fireEvent, render, waitFor } from '@testing-library/react';
import * as React from 'react';
import renderer from 'react-test-renderer';
import { IconTooltip } from './IconTooltip';

const description = 'Sample tooltip description';
describe('Snapshot testing IconTooltip Component', () => {
  it('renders correctly with default size and color', () => {
    const tooltip = renderer.create(<IconTooltip description={description} />).toJSON();
    expect(tooltip).toMatchSnapshot();
  });

  it('renders correctly with Icon color', () => {
    const tooltip = renderer
      .create(<IconTooltip description={description} iconColor="white" />)
      .toJSON();
    expect(tooltip).toMatchSnapshot();
  });

  it('renders correctly with HtmlElement description', () => {
    const Description = () => (
      <div>
        Sale price is discounted by <b>5%</b>. This fee goes to the market creator and liquidity
        providers.
      </div>
    );
    const tooltip = renderer
      .create(<IconTooltip description={<Description />} iconColor="white" />)
      .toJSON();
    expect(tooltip).toMatchSnapshot();
  });
});

describe('Element testing IconTooltip Component', () => {
  const setTooltipIsOpen = jest.fn();
  const useStateMock: any = (tooltipIsOpen: any) => [tooltipIsOpen, setTooltipIsOpen];

  it.skip('render correctly has been opened', async () => {
    const Description = () => (
      <div>
        Sale price is discounted by <b>5%</b>. This fee goes to the market creator and liquidity
        providers.
      </div>
    );
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    const { rerender, container, getByLabelText } = render(
      <IconTooltip description={<Description />} />,
    );
    rerender(<IconTooltip description={<Description />} />);
    await waitFor(() => {
      const buttonElement = getByLabelText(/tooltip/);
      // console.log(buttonElement);
      fireEvent.click(buttonElement);
    });
    expect(setTooltipIsOpen).toHaveBeenCalled();
    expect(container.querySelectorAll('b').length).toEqual(1);
  });
});
