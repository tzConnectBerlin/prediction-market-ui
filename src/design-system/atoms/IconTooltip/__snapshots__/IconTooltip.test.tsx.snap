// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing IconTooltip Component renders correctly with HtmlElement description 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  border-radius: 50%;
  background: rgba(1, 102, 255, 0.16);
  height: 0.7em;
  width: 0.7em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

<button
  aria-label="tooltip"
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-0"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <div
    className="emotion-1"
  >
    <svg
      fill="white"
      height="0.7em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="0.7em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M232 235h48v137h-48zM232 140h48v48h-48z"
      />
    </svg>
  </div>
</button>
`;

exports[`Snapshot testing IconTooltip Component renders correctly with Icon color 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  border-radius: 50%;
  background: rgba(1, 102, 255, 0.16);
  height: 0.7em;
  width: 0.7em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<button
  aria-label="tooltip"
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-0"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <div
    className="emotion-1"
  >
    <svg
      fill="white"
      height="0.7em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="0.7em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M232 235h48v137h-48zM232 140h48v48h-48z"
      />
    </svg>
  </div>
  <span
    className="MuiTouchRipple-root emotion-2"
  />
</button>
`;

exports[`Snapshot testing IconTooltip Component renders correctly with default size and color 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  border-radius: 50%;
  background: rgba(1, 102, 255, 0.16);
  height: 0.7em;
  width: 0.7em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

<button
  aria-label="tooltip"
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-0"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <div
    className="emotion-1"
  >
    <svg
      fill="#1976d2"
      height="0.7em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="0.7em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M232 235h48v137h-48zM232 140h48v48h-48z"
      />
    </svg>
  </div>
</button>
`;
